File: backend/controllers/authController.js
const bcrypt = require("bcryptjs");
const jwt = require("jsonwebtoken");
const db = require("../config/db");

const generateToken = (user) =>
  jwt.sign(
    { id: user.id, role: user.role, name: user.name },
    process.env.JWT_SECRET,
    { expiresIn: process.env.JWT_EXPIRES_IN || "7d" },
  );

// POST /api/auth/register
exports.register = async (req, res) => {
  try {
    const { name, email, password, role, phone, location, craft_type, bio } =
      req.body;

    if (!name || !email || !password || !role)
      return res.status(400).json({ message: "Missing required fields" });

    if (!["client", "artisan"].includes(role))
      return res.status(400).json({ message: "Invalid role" });

    // MySQL: const [existing] = await db.query("SELECT id FROM users WHERE email = ?", [email]);
    const existingResult = await db.query(
      "SELECT id FROM users WHERE email = $1",
      [email],
    );
    if (existingResult.rows.length)
      return res.status(409).json({ message: "Email already exists" });

    const hashed = await bcrypt.hash(password, 10);
    const is_approved = role === "client" ? 1 : 0;

    // MySQL: const [result] = await db.query("INSERT INTO users (...) VALUES (?,?,?,?,?,?,?)", [...]);
    const result = await db.query(
      "INSERT INTO users (name, email, password, role, phone, location, is_approved) VALUES ($1,$2,$3,$4,$5,$6,$7) RETURNING id",
      [name, email, hashed, role, phone || null, location || null, is_approved],
    );
    const newUserId = result.rows[0].id;

    if (role === "artisan") {
      await db.query(
        "INSERT INTO artisan_profiles (user_id, craft_type, bio) VALUES ($1,$2,$3)",
        [newUserId, craft_type || null, bio || null],
      );
    }

    const user = { id: newUserId, name, email, role };
    res.status(201).json({
      message:
        role === "artisan"
          ? "Registration submitted, awaiting approval"
          : "Registration successful",
      token: generateToken(user),
      user: { id: user.id, name, email, role, is_approved },
    });
  } catch (err) {
    console.log(err);
    res.status(500).json({ message: "Server error", error: err.message });
  }
};

// POST /api/auth/login
exports.login = async (req, res) => {
  try {
    const { email, password } = req.body;
    if (!email || !password)
      return res.status(400).json({ message: "Email and password required" });

    // MySQL: const [rows] = await db.query("SELECT * FROM users WHERE email = ?", [email]);
    const result = await db.query("SELECT * FROM users WHERE email = $1", [
      email,
    ]);
    if (!result.rows.length)
      return res.status(401).json({ message: "Invalid credentials" });

    const user = result.rows[0];
    const valid = await bcrypt.compare(password, user.password);
    if (!valid) return res.status(401).json({ message: "Invalid credentials" });

    if (user.role === "artisan" && !user.is_approved)
      return res.status(403).json({ message: "Account pending approval" });

    let profile = null;
    if (user.role === "artisan") {
      const profileResult = await db.query(
        "SELECT * FROM artisan_profiles WHERE user_id = $1",
        [user.id],
      );
      profile = profileResult.rows[0] || null;
    }

    res.json({
      token: generateToken(user),
      user: {
        id: user.id,
        name: user.name,
        email: user.email,
        role: user.role,
        avatar: user.avatar,
        location: user.location,
        is_approved: user.is_approved,
        profile,
      },
    });
  } catch (err) {
    res.status(500).json({ message: "Server error", error: err.message });
  }
};

// GET /api/auth/me
exports.me = async (req, res) => {
  try {
    const result = await db.query(
      "SELECT id, name, email, role, phone, location, avatar, is_approved FROM users WHERE id = $1",
      [req.user.id],
    );
    if (!result.rows.length)
      return res.status(404).json({ message: "User not found" });
    res.json({ user: result.rows[0] });
  } catch (err) {
    res.status(500).json({ message: "Server error", error: err.message });
  }
};

// PATCH /api/auth/profile
exports.updateProfile = async (req, res) => {
  try {
    const { name, phone, location, bio, craft_type } = req.body;
    await db.query(
      "UPDATE users SET name = $1, phone = $2, location = $3 WHERE id = $4",
      [name, phone || null, location || null, req.user.id],
    );
    if (req.user.role === "artisan") {
      await db.query(
        "UPDATE artisan_profiles SET bio = $1, craft_type = $2 WHERE user_id = $3",
        [bio || null, craft_type || null, req.user.id],
      );
    }
    res.json({ message: "Profile updated" });
  } catch (err) {
    res.status(500).json({ message: "Server error", error: err.message });
  }
};

File: flutter_app/lib/controllers/auth_controller.dart
import 'package:get/get.dart';
import '../core/services/api_service.dart';
import '../core/services/auth_service.dart';
import '../core/constants/app_constants.dart';

class AuthController extends GetxController {
  final _api = Get.find<ApiService>();
  final _auth = Get.find<AuthService>();

  final isLoading = false.obs;
  final errorMsg = ''.obs;

  Future<void> login(String email, String password) async {
    isLoading.value = true;
    errorMsg.value = '';
    try {
      final res = await _api
          .post('/auth/login', data: {'email': email, 'password': password});
      await _auth.saveSession(res.data['token'], res.data['user']);
      _navigateByRole(res.data['user']['role']);
    } catch (e) {
      errorMsg.value = _extractError(e);
    } finally {
      isLoading.value = false;
    }
  }

  Future<void> register({
    required String name,
    required String email,
    required String password,
    required String role,
    String? phone,
    String? location,
    String? craftType,
    String? bio,
  }) async {
    isLoading.value = true;
    errorMsg.value = '';
    try {
      final res = await _api.post('/auth/register', data: {
        'name': name,
        'email': email,
        'password': password,
        'role': role,
        'phone': phone,
        'location': location,
        'craft_type': craftType,
        'bio': bio,
      });
      await _auth.saveSession(res.data['token'], res.data['user']);
      if (role == 'artisan') {
        Get.snackbar('تم التسجيل', 'حسابك قيد المراجعة من الإدارة',
            snackPosition: SnackPosition.BOTTOM);
        Get.offAllNamed(AppRoutes.login);
      } else {
        _navigateByRole(role);
      }
    } catch (e) {
      errorMsg.value = _extractError(e);
    } finally {
      isLoading.value = false;
    }
  }

  void _navigateByRole(String role) {
    switch (role) {
      case 'client':
        Get.offAllNamed(AppRoutes.clientHome);
        break;
      case 'artisan':
        Get.offAllNamed(AppRoutes.artisanDashboard);
        break;
      case 'admin':
        Get.offAllNamed(AppRoutes.adminDashboard);
        break;
    }
  }

  Future<void> logout() async {
    await _auth.logout();
    Get.offAllNamed(AppRoutes.login);
  }

  String _extractError(dynamic e) {
    try {
      return e.response?.data['message'] ?? 'حدث خطأ، حاول مجدداً';
    } catch (_) {
      return 'تعذر الاتصال بالخادم';
    }
  }
}
